src/resource_reader.rs

```diff
@@ -13,14 +17,39 @@ pub struct FileRead {
 
 pub struct ResourceReader {
     runtime: TargetRuntime,
+    // Cache file content, path->content
+    cache: Arc<Mutex<HashMap<String, String>>>,
 }
 
 impl ResourceReader {
-    pub fn init(runtime: TargetRuntime) -> Self {
-        Self { runtime }
+    pub fn init(runtime: TargetRuntime, cache: Arc<Mutex<HashMap<String, String>>>) -> Self {
```

Comment:
```suggestion
    pub fn init(runtime: TargetRuntime, cache: bool) -> Self {
```

If cache is true, we should initialize it inside of `init`

Reply:
>And there is another issue, the first reading is called by ConfigReader::resource_reader, the second is called by ConfigReader::proto_reader, they are different object. So we need to get the HashMap from outside.

The cache should not be shared. If the cache needs to be shared, we should do that by sharing the ResourceReader.

Reply:
I think that's just an additional unwrapping that makes the code difficult to read. You can create two resource readers
```rust
struct ResourceReader<A>(A);

impl<A> ResourceReader<A> {
  fn direct(runtime: TargetRuntime) -> ResourceReader<Direct>
  fn cached(runtime: TargetRuntime) -> ResourceReader<Cache>
}

// Reads the files directly without cache
Direct {runtime: TargetRuntime }

// Reads it and maintains a cache
pub Cache {
  reader: Direct,
  cache: Arc<Mutex<HashMap<String, FileRead>>
}

```

Can we do something like above? We will know at compile time, if the reader is caching or not.

Reply:
It makes the code a lot more modular, type-safe and easy to test. The current implementation has a lot of checks before reading, increasing the potential